ballerina/auth_desugar.bal

```diff
@@ -0,0 +1,162 @@
+        }
+        if result is PermissionDeniedError {
+            return result;
+        } else if result is boolean && result {
+            return;
+        }
+    }
+    return error UnauthenticatedError("Failed to authenticate client");
+}
+
+isolated function authenticateWithFileUserStore(FileUserStoreConfigWithScopes config, map<string|string[]> headers)
+ returns PermissionDeniedError|boolean? {
+    ListenerFileUserStoreBasicAuthHandler handler = new(config.fileUserStoreConfig);
+    auth:UserDetails|UnauthenticatedError authn = handler.authenticate(headers);
+    string|string[]? scopes = config?.scopes;
+    if authn is auth:UserDetails {
+        if scopes is string|string[] {
+            PermissionDeniedError? authz = handler.authorize(authn, scopes);
+            if authz is PermissionDeniedError {
+                return authz;
+            }
+            return true;
+        }
+        return true;
+    }
+}
+
+isolated function authenticateWithLdapUserStore(LdapUserStoreConfigWithScopes config, map<string|string[]> headers)
+ returns PermissionDeniedError|boolean? {
+    ListenerLdapUserStoreBasicAuthHandler handler = new(config.ldapUserStoreConfig);
+    auth:UserDetails|UnauthenticatedError authn = handler->authenticate(headers);
+    string|string[]? scopes = config?.scopes;
+    if authn is auth:UserDetails {
+        if scopes is string|string[] {
+            PermissionDeniedError? authz = handler->authorize(authn, scopes);
+            if authz is PermissionDeniedError {
+                return authz;
+            }
+            return true;
+        }
+        return true;
+    }
+}
+
+isolated function authenticateWithJwtValidator(JwtValidatorConfigWithScopes config, map<string|string[]> headers)
+ returns PermissionDeniedError|boolean? {
+    ListenerJwtAuthHandler handler = new(config.jwtValidatorConfig);
+    jwt:Payload|UnauthenticatedError authn = handler.authenticate(headers);
+    string|string[]? scopes = config?.scopes;
+    if authn is jwt:Payload {
+        if scopes is string|string[] {
+            PermissionDeniedError? authz = handler.authorize(authn, scopes);
+            if authz is PermissionDeniedError {
+                return authz;
+            }
+            return true;
+        }
+        return true;
+    }
+}
+
+isolated function authenticateWithOAuth2IntrospectionConfig(OAuth2IntrospectionConfigWithScopes config, map<string|string[]> headers)
+ returns PermissionDeniedError|boolean? {
+    // Here, config is OAuth2IntrospectionConfigWithScopes
+    ListenerOAuth2Handler handler = new(config.oauth2IntrospectionConfig);
+    oauth2:IntrospectionResponse|UnauthenticatedError|PermissionDeniedError auth = handler->authorize(headers, config?.scopes);
+    if auth is oauth2:IntrospectionResponse {
+        return true;
+    } else if auth is PermissionDeniedError {
+        return auth;
+    }
+}
+
+isolated function getListenerAuthConfig(Service serviceRef)
```

Comment:
We can remove this function completely, and call the `getServiceAuthConfig` function directly from the parent function. This was used in other modules because we need to check both service and resource configurations and prioritize resource configs.

Reply:
Changed in https://github.com/ballerina-platform/module-ballerina-grpc/pull/343/commits/c156f03ddbc865d79ce4d3d3faf6183ef8016204